pkg/kudoctl/packages/resolver/resolver_incluster.go

```diff
@@ -0,0 +1,43 @@
+package resolver
+
+import (
+	"fmt"
+
+	kudoapi "github.com/kudobuilder/kudo/pkg/apis/kudo/v1beta1"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/packages"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/packages/convert"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/util/kudo"
+)
+
+// InClusterResolver resolves packages that are already installed in the cluster on the client-side. Note, that unlike
+// other resolvers, the resulting 'packages.Package' struct does not contain package 'packages.Files' (we don't have
+// the original files).
+type InClusterResolver struct {
+	c  *kudo.Client
+	ns string
+}
+
+func (r InClusterResolver) Resolve(name string, appVersion string, operatorVersion string) (*packages.Package, error) {
```

Comment:
I'm not really happy that the `Resolve` function from both `InClusterResolver` do not have the same behaviour as the ones for repos.

- They fully ignore the given `appVersion`
- They *require* an operatorVersion and don't return an error if none if given

I think in the end they should: 
- Look through all installed OVs
- If an OV-Version is given, filter by OV-Version
- If an App-Version is given, filter by App-Version
- Sort by the same order as the repo-Resolver
- Return the newest OV

Reply:
All the above makes sense for a repo but when in-cluster, _operator name + version_ is enough to uniquely identify any operator. So, there is no need to e.g. `Look through all installed OVs` - the one that we need is called e.g. `zookeeper-0.3.0` and there can _only be one of them per namespace._ Actually, the install command could look like `kudo install default/zookeeper-0.3.0 ...` but that doesn't look very kudoctl-like. 

We could demand setting an `appVersion` and compare it with the one used in `OV` but that would be only a sanity check.

Reply:
I don't agree that we should *require* an OV to be specified here. We don't require it for the installation from a repo, why should we require it when creating the instance from an installed Operator? And at that point it's only a very small step to replicate the full behaviour of installation from a repository.

Reply:
You're right - given the fact, that we'll move towards the versioning model described in the KEP-19 eventually, this makes sense.

Reply:
btw, I've mostly implemented that logic in the ov-rename-branch, and it's mostly reusable. Doesn't have a PR open yet though

Reply:
> And at that point it's only a very small step to replicate the full behavior of installation from a repository.

I slept over it and I think, this is not the same UX. Let's say a user wants to install a zookeeper instance from an existing in-cluster `OV`. What's likely to be the first step? Probably checking what's there:

```
❯ ./bin/kubectl-kudo get operatorversions
List of current installed operatorversions in namespace "default":
.
├── child-0.0.1
├── dummy-0.1.0
├── first-operator-0.2.0
├── flink-0.2.1
├── flink-demo-0.1.5
├── kafka-1.2.0
├── parent-0.1.0
└── zookeeper-0.3.0
```

Maybe it is worth checking the `appVersion`:
```
❯ k get operatorversions.kudo.dev zookeeper-0.3.0 -o yaml | rg appVersion
  appVersion: 3.4.14
```
though, if it was pre-installed by the sysadmins (or the same user) then it's already known.

So, now that the user found the `OV`, she wants to install **from that exact `OV`** e.g.:

```
❯ kudo install zookeeper-0.3.0 --in-cluster # later zookeeper-0.3.0_3.4.14 or similar
```

but something like:
```
kudo install ... zookeeper --operator-version 0.3.0 --app-version 3.4.14 --in-cluster
```
doesn't have the same semantic. It is **not** the same as installing from a repo.

Reply:
Hmmmm... I'm not sure why it doesn't have the same semantic?

Having said that, i'm ok with
```
kudo install zookeeper-3.4.14-0.3.0 --in-cluster
```

So, specifying a full OV name should work as you describe.

But, on the other hand, I'd expect
```
kudo install zookeeper --in-cluster
```
To work the same as with a repo - It installs an instance of the newest OV that's installed in the cluster. 

I'm not sure about the `--operator-version` and `--app-version`. I would expect them to work the same a with a repo - filter the available versions by operator version or app version.

Reply:
I don't know whose side I am on to be honest but I feel like we're lacking some real user stories for this. In my mind when I always thought about installing instance I always thought about doing it to one hand-picked OV because I know that I want an instance of zookeeper 3.1 not zookeeper 2.2. I can't really imagine I would leave that to "faith" and just be ok with ANY zookeeper outside some testing environment or localhost.

But that said, I don't really know how people use it "in the wild" or intend to use it. Should we maybe implement install by a particular OV right now but don't close the doors to the repo resolve story Andreas is mentioning and implement that as well when someone requests it?

Reply:
Yeah, I wouldn't mind that - I think installing a very specific OV would be a good first use case, i.e. `kudo install zookeeper-3.4.14-0.3.0 --in-cluster` would work, `kudo install zookeeper --in-cluster` would fail with an error saying you need to give a specific version. `--operator-version` and `--app-version` would be forbidden in this scenario.


Apart from this: I usually would rather specify the app version than the operator version:
`kudo install zookeeper --app-version 3.4.14` makes a lot of sense, I want the latest operator for this app version.

Reply:
I don't think, _mixing operator names with resource names_ is a good idea. It's either-or. Otherwise, we have to parse e.g. `zookeeper-0.3.0` or `zookeeper-0.3.0_3.4.14` to see whether the user meant a resource or operator name. 

Anyway, I'm myself not particularly convinced in any of the ways and we don't have real user stories. Anyway, the last commit makes `--in-cluster` option behave the same way as the repo resolver  (same operator matching rules).